x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/CategorizeTextAggregator.java

```diff
@@ -0,0 +1,245 @@
+
+    private final TermsAggregator.BucketCountThresholds bucketCountThresholds;
+    private final SourceLookup sourceLookup;
+    private final MappedFieldType fieldType;
+    private final CategorizationAnalyzer analyzer;
+    private final String sourceFieldName;
+    private ObjectArray<CategorizationTokenTree> categorizers;
+    private final int maxUniqueTokens;
+    private final int maxMatchTokens;
+    private final int similarityThreshold;
+    private final LongKeyedBucketOrds bucketOrds;
+    private final CategorizationBytesRefHash bytesRefHash;
+
+    protected CategorizeTextAggregator(
+        String name,
+        AggregatorFactories factories,
+        AggregationContext context,
+        Aggregator parent,
+        String sourceFieldName,
+        MappedFieldType fieldType,
+        TermsAggregator.BucketCountThresholds bucketCountThresholds,
+        int maxUniqueTokens,
+        int maxMatchTokens,
+        int similarityThreshold,
+        CategorizationAnalyzerConfig categorizationAnalyzerConfig,
+        Map<String, Object> metadata
+    ) throws IOException {
+        super(name, factories, context, parent, metadata);
+        this.sourceLookup = context.lookup().source();
+        this.sourceFieldName = sourceFieldName;
+        this.fieldType = fieldType;
+        CategorizationAnalyzerConfig analyzerConfig = Optional.ofNullable(categorizationAnalyzerConfig)
+            .orElse(CategorizationAnalyzerConfig.buildStandardCategorizationAnalyzer(Collections.emptyList()));
+        String analyzer = analyzerConfig.getAnalyzer();
+        final boolean shouldClose;
+        final Analyzer innerAnalyzer;
+        if (analyzer != null) {
+            Analyzer globalAnalyzer = context.getNamedAnalyzer(analyzer);
+            if (globalAnalyzer == null) {
+                throw new IllegalArgumentException("Failed to find global analyzer [" + analyzer + "]");
+            }
+            innerAnalyzer = globalAnalyzer;
+            shouldClose = false;
+        } else {
+            innerAnalyzer = context.buildCustomAnalyzer(
+                context.getIndexSettings(),
+                false,
+                analyzerConfig.getTokenizer(),
+                analyzerConfig.getCharFilters(),
+                analyzerConfig.getTokenFilters()
+            );
+            shouldClose = true;
+        }
+        this.analyzer = new CategorizationAnalyzer(innerAnalyzer, shouldClose);
+        this.categorizers = bigArrays().newObjectArray(1);
+        this.maxUniqueTokens = maxUniqueTokens;
+        this.maxMatchTokens = maxMatchTokens;
+        this.similarityThreshold = similarityThreshold;
+        this.bucketOrds = LongKeyedBucketOrds.build(bigArrays(), CardinalityUpperBound.MANY);
+        this.bucketCountThresholds = bucketCountThresholds;
+        this.bytesRefHash = new CategorizationBytesRefHash(new BytesRefHash(2048, bigArrays()));
+    }
+
+    @Override
+    protected void doClose() {
+        super.doClose();
```

Comment:
`Releasables.close` has the lovely side effect of trying to close everything even if something fails. And it throws an unchecked exception.